python/ray/autoscaler/_private/kuberay/node_provider.py

```diff
@@ -0,0 +1,234 @@
+# back to the autoscaler. The waiting period is typically small, on the order
+# of a few seconds. We make sure that only one such modification is in process
+# by serializing all modification operations with a lock in the NodeProvider.
+
+
+# Note: Log handlers set up in autoscaling monitor entrypoint.
+logger = logging.getLogger(__name__)
+
+provider_exists = False
+
+
+def to_label_selector(tags):
+    label_selector = ""
+    for k, v in tags.items():
+        if label_selector != "":
+            label_selector += ","
+        label_selector += "{}={}".format(k, v)
+    return label_selector
+
+def status_tag(pod: Dict[str, Any]):
+    if "containerStatuses" not in pod["status"]:
+        return "pending"
+    
+    state = pod["status"]["containerStatuses"][0]["state"]
+
+    logger.info("status_tag: state = {}".format(state))
+
+    if "pending" in state:
+        return "pending"
+    if "running" in state:
+        return "up-to-date"
+    if "waiting" in state:
+        return "waiting"
+    if "terminated" in state:
+        return "update-failed"
+    raise ValueError("Unexpected container state.")
+
+def make_node_tags(labels: Dict[str, str], status_tag: str):
+    "Convert Kuberay labels to Ray autoscaler tags."
+    tags = {"ray-node-status": status_tag}
+
+    if labels["ray.io/node-type"] == "head":
+        tags["ray-node-type"] = "head"
+        tags["ray-user-node-type"] = "head-group"
+    else:
+        tags["ray-node-type"] = "worker"
+        tags["ray-user-node-type"] = labels["ray.io/group"]
+
+    return tags
+
+
+class KuberayNodeProvider(NodeProvider):  # type: ignore
+    def __init__(
+        self,
+        provider_config: Dict[str, Any],
+        cluster_name: str,
+        _allow_multiple: bool = False,
+    ):
+        logger.info("Creating ClientNodeProvider.")
+        self.namespace = provider_config["namespace"]
+        self.cluster_name = cluster_name
+        self._lock = threading.RLock()
+
+        with open("/var/run/secrets/kubernetes.io/serviceaccount/token") as secret:
+            token = secret.read()
+
+        self.headers = {
+            "Authorization": "Bearer " + token,
+        }
+        self.verify = "/var/run/secrets/kubernetes.io/serviceaccount/ca.crt"
+
+        # Disallow multiple node providers, unless explicitly allowed for testing.
+        global provider_exists
+        if not _allow_multiple:
+            assert (
+                not provider_exists
+            ), "Only one ClientNodeProvider allowed per process."
+        assert (
+            provider_config.get("disable_node_updaters", False) is True
+        ), "Must disable node updaters to use ClientNodeProvider."
+        provider_exists = True
+
+        super().__init__(provider_config, cluster_name)
+
+    def _get(self, path):
+        result = requests.get("https://kubernetes.default:443" + path, headers=self.headers, verify=self.verify)
```

Comment:
Trying to understand why not to use kubernetes python client? Do you think that's the dependency not easy to manage?

Reply:
I did the initial implementation with the kubernetes python client, but it had a bug where PATCH was not implemented correctly (I believe this is https://github.com/kubernetes-client/python/pull/959). Given there is such an egregious bug I didn't trust the rest of the client and this implementation is simple enough and should be super robust. Also getting away with less dependencies is always nice and the K8s REST API is supported well. So I'm happy to keep it the way it is but also happy to switch it out in the future if we prefer (I do think @DmitriGekhtman has a workaround for this bug now, basically using the underlying REST client directly).